### PR TITLE
Accept receipts on events related to the root event.

### PR DESCRIPTION
**Pull request by clokep**
See the discussion in #14168, this is a minor spec-compliance fix for how clients expect receipts to work.

Builds on #14163 

More tests at matrix-org/complement#501.

### CODE REVIEW

File: synapse/rest/client/receipts.py [hunk 1]
```diff
@@ -109,6 +109,46 @@ async def on_POST(
...
+    async def _is_valid_thread_id(self, event_id: str, thread_id: str) -> bool:
+        """
+        The thread ID provided must relate (in a vague sense) to the event ID.
+
+        We check this to ensure clients aren't sending bogus receipts.
+
+        A thread ID is considered valid if:
+
+        1. The event has a thread relation which matches the thread ID.
+        2. The event has children events which form a thread relation (i.e. the
```
Comment (DMRobertson) [inline synapse/rest/client/receipts.py:121]: Some nitpicks here.

- Does "children events" mean direct children or arbitrary descendants?
- Why "events" plural here? Presumably the existence of at least one such child events is sufficient?
- "form a thread relation" presumably means that they are marked as relating (via m.thread) _to the event in question_.

File: synapse/rest/client/receipts.py [hunk 2]
```diff
@@ -109,6 +109,46 @@ async def on_POST(
...
+           event is a thread root).
+        2. The event is related to an event (recursively) which satisfies 1 or 2.
```
Comment (DMRobertson) [inline synapse/rest/client/receipts.py:123]: I think you want `3.` here.

Is any relation permitted here? There's potential to confuse this with "relate" on +114.

https://github.com/matrix-org/synapse/issues/14168#issuecomment-1277670830 talks about a limit to recursion that we go on to implement below and in the next commit. Should we mention that here?

File: synapse/rest/client/receipts.py [hunk 3]
```diff
@@ -109,6 +109,46 @@ async def on_POST(
...
+
+        Given the following DAG:
+
+            A <---[m.thread]-- B <--[m.annotation]-- C
+            ^
+            |--[m.reference]-- D <--[m.annotation]-- E
```
Comment (DMRobertson) [inline synapse/rest/client/receipts.py:129]: What is m.reference? I can't see its meaning described in https://spec.matrix.org/v1.4/client-server-api/

File: synapse/rest/client/receipts.py [hunk 4]
```diff
@@ -109,6 +109,46 @@ async def on_POST(
...
+
+        It is valid to send a receipt for thread A on A, B, C, D, or E.
+
+        It is valid to send a receipt for the main thread on A, D, and E.
```
Comment (DMRobertson) [inline synapse/rest/client/receipts.py:133]: By "main thread" do we mean the top-level room timeline? We call it MAIN_TIMELINE below---maybe call it "main timeline" here?

File: synapse/rest/client/receipts.py [hunk 5]
```diff
@@ -109,6 +109,46 @@ async def on_POST(
...
+
+        Args:
+            event_id: The event ID to check.
+            thread_id: The thread ID the event is potentially part of.
+
+        Returns:
+            True if the event belongs to the given thread.
```
Comment (DMRobertson) [inline synapse/rest/client/receipts.py:140]: Maybe overkill?

```suggestion
            True if the event belongs to the given thread, otherwise False.
```

File: synapse/storage/databases/main/relations.py [hunk 6]
```diff
@@ -946,13 +946,19 @@ async def get_thread_id(self, event_id: str) -> str:
         Get the thread ID for an event. This considers multi-level relations,
         e.g. an annotation to an event which is part of a thread.
 
+        It only searches up the relations tree, i.e. it only considers events
+        which are the parent of the given event.
```
Comment (DMRobertson) [inline synapse/storage/databases/main/relations.py:950]: What does "parent" mean here? Is P a parent of E if P belongs to E's list of `prev_events`? Or do you mean that E marks itself as relating to P?

File: synapse/storage/databases/main/relations.py [hunk 7]
```diff
@@ -978,6 +984,59 @@ def _get_thread_id(txn: LoggingTransaction) -> str:
...
+                SELECT event_id, relates_to_id, relation_type, 0 depth
+                FROM event_relations
+                WHERE event_id = ?
+                UNION SELECT e.event_id, e.relates_to_id, e.relation_type, depth + 1
+                FROM event_relations e
+                INNER JOIN related_events r ON r.relates_to_id = e.event_id
+                WHERE depth <= 3
+            ) SELECT relates_to_id FROM related_events ORDER BY depth DESC LIMIT 1
+        ), ?) AND relation_type = 'm.thread' LIMIT 1;
+        """
```
Comment (DMRobertson) [inline synapse/storage/databases/main/relations.py:1032]: 1. I'm happy that the WITH RECURSIVE expression recurses 3 steps up the relations tree.
2. We then SELECT the `relates_to_id` of the last thing we saw in that WITH RECURSIVE expression.
3. We then look for a relation of type m.thread pointing at the event from 2.

Does this work if event A relates to some even older event X? E.g. suppose that A was sent in reply to X. Wouldn't that mean that in point 2 above we select the event id of X?

File: tests/storage/test_relations.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+            F <--[m.annotation]-- G
+
+        """
+        self._main_store = self.hs.get_datastores().main
+
+        self._create_relation("A", "B", "m.thread")
+        self._create_relation("B", "C", "m.annotation")
+        self._create_relation("A", "D", "m.reference")
+        self._create_relation("A", "E", "m.annotation")
```
Comment (DMRobertson) [inline tests/storage/test_relations.py:41]: I think A should be D here, judging by the picture?

File: synapse/rest/client/receipts.py [hunk 9]
```diff
@@ -109,6 +109,46 @@ async def on_POST(
 
         return 200, {}
 
+    async def _is_valid_thread_id(self, event_id: str, thread_id: str) -> bool:
```
Comment (DMRobertson) [inline synapse/rest/client/receipts.py:112]: Maybe call this `_is_event_in_thread` or `_event_in_thread`? We're not doing any validation here, if you see what I mean.

(Sorry for bikeshedding.)

File: synapse/storage/databases/main/relations.py [hunk 10]
```diff
@@ -978,6 +984,59 @@ def _get_thread_id(txn: LoggingTransaction) -> str:
 
         return await self.db_pool.runInteraction("get_thread_id", _get_thread_id)
 
+    @cached()
```
Comment (DMRobertson) [inline synapse/storage/databases/main/relations.py:991]: Is this safe to cache? Sanity checks:

- is it possible for a given event to belong to more than one thread?
- is it possible that an event could belong to some thread at one moment in time, then later not belong to a thread (e.g. due to state resolution invalidating something that was previously accepted)

**Review by DMRobertson — COMMENTED** [review]
Sorry Patrick, I have erred on the side of caution and pedantry here. I hope you find my comments useful rather than irritating.

File: synapse/rest/client/receipts.py (hunk 3 above)
Comment (clokep) [reply synapse/rest/client/receipts.py:129]: See matrix-org/matrix-spec#1206. (Note that annotations aren't even accepted: matrix-org/matrix-spec-proposals#2677, but they're in a weird spot and generally considered supported...I think having it here is fine.)

File: synapse/rest/client/receipts.py (hunk 4 above)
Comment (clokep) [reply synapse/rest/client/receipts.py:133]: Ah, that was a typo. Should be main timeline! Thank you.

File: synapse/storage/databases/main/relations.py (hunk 10 above)
Comment (clokep) [reply synapse/storage/databases/main/relations.py:991]: > * is it possible for a given event to belong to more than one thread?

No.

> * is it possible that an event could belong to some thread at one moment in time, then later not belong to a thread (e.g. due to state resolution invalidating something that was previously accepted)

Only if the event is redacted, which is handled via invalidation in this PR.

File: tests/storage/test_relations.py (hunk 8 above)
Comment (DMRobertson) [reply tests/storage/test_relations.py:41]: (on line 41)

File: synapse/storage/databases/main/relations.py (hunk 10 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/relations.py:991]: Thanks. The key thing I hadn't realised is that an event can only cite at most one other event that it relates to. This makes the relationships graph a tree.

File: synapse/rest/client/receipts.py (hunk 1 above)
Comment (clokep) [reply synapse/rest/client/receipts.py:120]: > Does "children events" mean direct children or arbitrary descendants?

A thread relation cannot relate to an event which has a relation, so it means direct child.

> Why "events" plural here? Presumably the existence of at least one such child events is sufficient?

Sloppiness, one child is sufficient.

> "form a thread relation" presumably means that they are marked as relating (via m.thread) _to the event in question_.

Yes. It is trying to say that event `A` has another event `B` with an `m.relates_to` of type `m.thread` pointed at `A`.

File: synapse/storage/databases/main/relations.py (hunk 6 above)
Comment (clokep) [reply synapse/storage/databases/main/relations.py:950]: It is in reference to relations. Note that [the spec](https://spec.matrix.org/v1.4/client-server-api/#forming-relationships-between-events) uses the parent/child terminology. Do you have suggestions for clearer terms?

File: synapse/rest/client/receipts.py (hunk 2 above)
Comment (clokep) [reply synapse/rest/client/receipts.py:122]: > Is any relation permitted here? There's potential to confuse this with "relate" on +114.

Yes (as long as it abides by the rules of that relation).

> https://github.com/matrix-org/synapse/issues/14168#issuecomment-1277670830 talks about a limit to recursion that we go on to implement below and in the next commit. Should we mention that here?

Maybe? It is somewhat an implementation detail, but likely worth mentioning. Maybe I'll leave it a bit vague and say "recursively, up to a point" or something?

File: synapse/storage/databases/main/relations.py (hunk 7 above)
Comment (clokep) [reply synapse/storage/databases/main/relations.py]: I'm not really sure what you're asking. I think you're asking is what would happen if you have:

```mermaid
flowchart RL
    A-.->|m.reference|X
    B-.->|m.thread|A
```

This is an invalid structure -- a thread event cannot apply to an event which has a relation itself (i.e. the relation between `B` and `A` is invalid).

Comment (clokep) [reply synapse/storage/databases/main/relations.py]: (I should probably include a bit more of a definition about how this query works though...)

File: synapse/storage/databases/main/relations.py (hunk 6 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/relations.py:950]: Thoughts:
- parent/child
- relatee/relator (urgh)
- stimulus/reaction
- relation target/relation source

None of these are great though. Possibly "relation-parent"? (I've used the phrasing "auth-parent" elsewhere in reference to `auth_events`)

Another idea might be to rephrase this to avoid references ancestry, e.g.

```suggestion
        It only searches up the relations tree, i.e. it only looks for events
        that the given event cites in its `m.relates_to` content field.
```

or 

```suggestion
        It only searches up the relations tree, i.e. it only considers events
        which the given event declares itself as being related to.
```

File: synapse/rest/client/receipts.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/rest/client/receipts.py:120]: I think we have are working with different meanings of "child". I understand "A is a child of B" to mean "A cites B as a prev_event". But it sounds here that you understand "A is a child of B" to mean "A declares itself as being `m.relates_to` B via `m.thread`". (See also the other comment thread about terminology).

Let me see if I can make a suggestion...

File: synapse/storage/databases/main/relations.py (hunk 7 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/relations.py]: > I think you're asking is what would happen if you have:

Yes, exactly this.

> This is an invalid structure 

Oh, well that makes life much easier! And I guess it means we can say that threads are closed under relations, which seems like a neat invariant.

File: synapse/rest/client/receipts.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/rest/client/receipts.py:120]: I'd phrase it like this, but I do lean towards the mathematical and the formal. Just a suggestion, feel free to ignore!

```
        The event must be related to the thread ID (in a vague sense) to ensure
        clients aren't sending bogus receipts.
        
        A thread ID is considered valid for a given event E if:
        
        1. E has a thread relation which matches the thread ID;
        2. some other event has a thread relation to E matching the thread ID; or
        3. E is recursively related (via any rel_types) to an event
           which satisfies 1 or 2.
```

**Review by DMRobertson — APPROVED** [review]
Thanks for bearing with me---I understand the threads data model (at the protocol level) much better now.

I think this looks good. The only outstanding comments are me fussing over phrasing. I've made some wording suggestions, but I'm happy to leave them to your discretion!

File: synapse/storage/databases/main/relations.py (hunk 6 above)
Comment (clokep) [reply synapse/storage/databases/main/relations.py:950]: Neither of those are really correct though, because it also looks at that recursively? I"ll try to play with it.

Comment (DMRobertson) [reply synapse/storage/databases/main/relations.py:950]: > Neither of those are really correct though, because it also looks at that recursively? I"ll try to play with it.

True, but I think the current wording also suffers from this.

**Review by DMRobertson — APPROVED** [review]
